src/SimpleSwap.sol

```diff
@@ -0,0 +1,164 @@
+        require(xIn > 0 && yIn > 0, "SimpleSwap: Can't provide 0 tokens");
+        x_0 += xIn;
+        y_0 += yIn;
+        zOut = xIn * yIn;
+        tokenX.safeTransferFrom(msg.sender, address(this), xIn);
+        tokenY.safeTransferFrom(msg.sender, address(this), yIn);
+        _mint(msg.sender, zOut);
+        emit LiquidityProvided(msg.sender, xIn, yIn, zOut);
+        }
+
+    /// @notice used to add liquidity to an existing pool
+    /// @param xIn the amount of token X that'll be added to the pool
+    /// @return zOut the amount of LP tokens that'll be minted to the user
+    function addLiquidity(uint256 xIn) external returns(uint256 zOut){
+        require(_totalSupply > 0, "SimpleSwap:Pool not intiated");
+        uint256 _x_0 = x_0;
+        uint256 _y_0 = y_0;
+        uint256 yIn = _y_0 * xIn / _x_0;
+        zOut = xIn * _totalSupply / _x_0;
+        x_0 += xIn;
+        y_0 += yIn;
+        tokenX.safeTransferFrom(msg.sender, address(this), xIn);
+        tokenY.safeTransferFrom(msg.sender, address(this), yIn);
+        _mint(msg.sender, zOut);
+        emit LiquidityProvided(msg.sender, xIn, yIn, zOut);
+    }
+
+    /// @notice used to remove tokens from the liquidity pool
+    /// @param zIn the amount of LP tokens to burn
+    /// @return xOut the amount of token X returned to the user 
+    /// @return yOut the amount of token Y returned to the user
+    function removeLiquidity(uint256 zIn) external returns(uint256 xOut, uint256 yOut) {
+        uint256 zOwnership = WDiv.wdiv(zIn, _totalSupply);
+        xOut = WMul.wmul(x_0, zOwnership);
+        yOut = WMul.wmul(y_0, zOwnership);
+        _burn(msg.sender, zIn);
+        tokenX.safeTransfer(msg.sender, xOut);
+        tokenY.safeTransfer(msg.sender, yOut);
+        emit LiquidityRemoved(msg.sender, xOut, yOut, zIn);
+    }
+
+    /// @notice used to swap an amount of token X for token Y
+    /// @param xIn the amount of token X being sold to the pool
+    /// @return yOut the amount of token Y being bought from the pool
+    function swapXForY(uint256 xIn) external returns(uint256 yOut){
+        yOut = _price(xIn, x_0, y_0);
+        x_0 += xIn;
+        y_0 -= yOut;
+        tokenX.safeTransferFrom(msg.sender, address(this), xIn);
+        tokenY.safeTransfer(msg.sender, yOut);
+        emit Swap(msg.sender, xIn, 0, 0, yOut);
+    }
+
+    /// @notice used to swap amount of token Y for token X
+    /// @param yIn the amount of token Y being sold to the pool
+    /// @param xOut the amount of token X being bought from the pool
+    function swapYForX(uint256 yIn) external returns(uint256 xOut){
+        xOut = _price(yIn, y_0, x_0);
+        y_0 += yIn;
+        x_0 -= xOut;
+        tokenY.safeTransferFrom(msg.sender, address(this), yIn);
+        tokenX.safeTransfer(msg.sender, xOut);
+        emit Swap(msg.sender, 0, yIn, xOut, 0);
+    }
+
+    function swapXforExactY(uint256 yOut) external returns(uint256 xIn){
```

Comment:
Could perhaps add natspec to some of these functions to unify the documentation but otherwise this file looks very clean.